Allow Nextflow tests workflow to self-fix with /fix-tests comment

nwiltsie opened the pull request:
# Description
This PR is a replacement for #28, with a (hopefully) smoother and more reviewable process. You can see the full capabilities in https://github.com/uclahs-cds/user-nwiltsie-pipeline/pull/11.

## User-Facing Flow for Broken Tests
Whenever a pull request is updated with a commit that causes the tests to fail, the bot will post a comment detailing the changes (helpfully provided by [jd](https://github.com/josephburnett/jd)):

<img width="597" alt="Screenshot 2024-03-22 at 11 38 27 AM" src="https://github.com/uclahs-cds/tool-Nextflow-action/assets/829731/cad3bd09-0fa7-4b45-8db2-5cd4c054de15">

The /fix-tests comment is inspired by https://github.com/peter-evans/slash-command-dispatch, but I'm not actually using it. The bot reacts with a 🚀 once the workflow starts to acknowledge the request. I also figured out how to use a separate token so that the tests will run on the fix commit itself - that eliminates the need for the suggested changes:

<img width="600" alt="Screenshot 2024-03-22 at 11 38 45 AM" src="https://github.com/uclahs-cds/tool-Nextflow-action/assets/829731/0a4ce45a-7713-47b3-b3cf-7e3d596ea15d">

## Edge Cases

### Needs Merge

A quirk of GitHub workflows is that they run on the _merge_ commit, not on the source branch commit. That's good, and what we want, but it can lead to strange cases like https://github.com/uclahs-cds/pipeline-call-gSV/pull/130 (specifically [this commit](https://github.com/uclahs-cds/pipeline-call-gSV/tree/ea19d9986a1d2610618dfa43b492e82189a3ed67)) where tests that did not exist in the branch were failing.

In cases like this the bot tells you how to resolve the issue:

<img width="597" alt="Screenshot 2024-03-22 at 11 38 13 AM" src="https://github.com/uclahs-cds/tool-Nextflow-action/assets/829731/e3943d58-8f2e-4080-b623-bff8c4f4529d">

### No Changes

If you request that the tests be fixed when there are no problems, the bot replies with some information:

<img width="596" alt="Screenshot 2024-03-22 at 11 38 01 AM" src="https://github.com/uclahs-cds/tool-Nextflow-action/assets/829731/cc06777b-8042-4f2f-9996-31bac9468c2e">


# Checklist
<!--- Please read each of the following items and confirm by replacing the [ ] with a [X] --->

- [x] This PR **does *NOT* contain** Protected Health Information [(PHI)](https://ohrpp.research.ucla.edu/hipaa/). A repo may ***need to be deleted*** if such data is uploaded. <br> Disclosing PHI is a ***major problem***[^1] - Even ***a small leak can be costly***[^2].
  
- [x] This PR **does *NOT* contain** germline genetic data[^3], RNA-Seq, DNA methylation, microbiome or other molecular data[^4].

[^1]: [UCLA Health reaches $7.5m settlement over 2015 breach of 4.5m patient records](https://healthitsecurity.com/news/ucla-health-reaches-7.5m-settlement-over-2015-breach-of-4.5m)
[^2]: [The average healthcare data breach costs $2.2 million, despite the majority of breaches releasing fewer than 500 records.](https://www.ponemon.org/local/upload/file/Sixth%20Annual%20Patient%20Privacy%20%26%20Data%20Security%20Report%20FINAL%206.pdf)
[^3]: [Genetic information is considered PHI.](https://www.genome.gov/about-genomics/policy-issues/Privacy#:~:text=In%202013%2C%20as%20required%20by,genetic%20information%20for%20underwriting%20purposes.)
  [Forensic assays can identify patients with as few as 21 SNPs](https://www.sciencedirect.com/science/article/pii/S1525157817305962)
[^4]: [RNA-Seq](https://www.nature.com/articles/ng.2248), [DNA methylation](https://ieeexplore.ieee.org/document/7958619), [microbiome](https://www.pnas.org/doi/pdf/10.1073/pnas.1423854112), or other molecular data can be used to predict genotypes (PHI) and reveal a patient's identity.


- [x] This PR **does *NOT* contain** other non-plain text files, such as: compressed files, images (*e.g.* `.png`, .`jpeg`), `.pdf`, `.RData`, `.xlsx`, `.doc`, `.ppt`, or other output files.

_&emsp; To automatically exclude such files using a [.gitignore](https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files) file, see [here](https://github.com/uclahs-cds/template-base/blob/main/.gitignore) for example._

- [x] I have read the [code review guidelines](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3187646/Code+Review+Guidelines) and the [code review best practice on GitHub check-list](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3189956/Code+Review+Best+Practice+on+GitHub+-+Check+List).

- [x] I have set up or verified the `main` branch protection rule following the [github standards](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3190380/GitHub+Standards#GitHubStandards-Branchprotectionrule) before opening this pull request.

- [x] The name of the branch is meaningful and well formatted following the [standards](https://uclahs-cds.atlassian.net/wiki/spaces/BOUTROSLAB/pages/3189956/Code+Review+Best+Practice+on+GitHub+-+Check+List), using [AD_username (or 5 letters of AD if AD is too long)]-[brief_description_of_branch].
  
- [x] I have added the major changes included in this pull request to the `CHANGELOG.md` under the next release version or unreleased, and updated the date.

yashpatel6 commented on .github/workflows/nextflow-restart.yml:
I may be missing it but is this workflow being used anywhere?

nwiltsie replied:
Nope! This is a holdover, I'll remove it.

nwiltsie replied:
Done with f6ea1dd.

yashpatel6 commented on .github/workflows/nextflow-tests.yml:
It would be worth trying to put this as `action.yml` under `run-nextflow-tests` so that it can be references without needing to go through `.github` with something like `uclahs-cds/tool-Nextflow-action/run-nextflow-tests@main`

nwiltsie replied:
They're not one-to-one equivalent - the key thing is that composite actions can't have multiple jobs, so the current flow of "discover, fan-out to parallelize jobs, fan-in for summary" scheme wouldn't work anymore.

As clunky as this calling `uclahs-cds/tool-Nextflow-action/.github/workflows/nextflow-tests.yml@main` is, I think it's a better choice than turning this into a composite action.

yashpatel6 replied:
Ah I see, that's fair!

yashpatel6 reviewed (commented):
Couple of comments:

yashpatel6 commented on .github/workflows/nextflow-tests.yml:
```diff
@@ -4,12 +4,99 @@ name: 'Nextflow tests'
+            let abort_early = false
+
+            if (apply_fixes) {
+              let bot_id = 41898282
```
Is this a randomly chosen number? I wonder if there's ever a possibility of the ID matching an actual user at some point

nwiltsie replied:
No, it's the user ID of the GitHub actions bot: https://api.github.com/users/github-actions%5Bbot%5D, so no other use should ever match it.

nwiltsie replied:
At some point I had a much fuller comment about this, but I stripped it at some point for some reason:

https://github.com/uclahs-cds/tool-Nextflow-action/blob/035c9423f0de304c609dbd5e8e0b0bdab445a715/.github/workflows/nextflow-tests.yml#L123-L130

yashpatel6 replied:
Got it, makes sense

aholmes replied:
I suggest leaving a comment like `// Actions bot ID - see https://api.github.com/users/github-actions%5Bbot%5D`

nwiltsie replied:
Done with 43877d5

yashpatel6 reviewed (approved):
I'm generally ok with this, anything else to add from @aholmes ?

aholmes commented on .github/workflows/nextflow-tests.yml:
```diff
@@ -4,12 +4,99 @@ name: 'Nextflow tests'
+            We need to determine if the incoming event was from a pull_request
+            or an issue_comment
+            */
+            let apply_fixes = context.eventName == 'issue_comment'
```
At the risk of starting a war, we should considering requiring semicolons. No change requested for now, just a thought for later.

nwiltsie replied:
Totally fair - I took a cue from the [docs](https://github.com/actions/github-script?tab=readme-ov-file#welcome-a-first-time-contributor) _not_ using them, but I do generally fall on the side of being explicit.

aholmes replied:
Gotcha - makes sense. A discussion for later, either way. :)

aholmes commented on .github/workflows/nextflow-tests.yml:
```diff
@@ -4,12 +4,100 @@ name: 'Nextflow tests'
+                }
+              }
+
+              if (!found_review) {
```
I suggest inverting if/else -
```javascript
if (found_review) {}
else {}
```
vs
```javascript
if (!found_review) {}
else {}
```

In my experience, reading the code tends to flow more naturally, requiring less cognitive load to parse the initial meaning. However, I know this can be subjective.

nwiltsie replied:
Done with 64f874d.

aholmes commented on .github/workflows/nextflow-tests.yml:
```diff
@@ -73,11 +169,257 @@ jobs:
+              needs.run.result != 'skipped'
+
+            var should_comment = tests_failed &&
+                !(needs.discover.outputs.apply_fixes == 'true')
```
Why `!(... == 'true')` rather than `(... != 'true')

nwiltsie replied:
I was so deep underwater at the time: https://github.com/uclahs-cds/tool-Nextflow-action/commit/0b118038cb31f072646584cf80636eca73a79dfd

I'll clean it up.

aholmes replied:
Been there. :)

nwiltsie replied:
Done with e7c93c7.

aholmes commented on .github/workflows/nextflow-tests.yml:
```diff
@@ -73,11 +169,257 @@ jobs:
+            var tests_failed = needs.run.result != 'success' &&
+              needs.run.result != 'skipped'
+
+            var should_comment = tests_failed &&
```
I suggest sticking to just `var` or `let` if it's appropriate for this code (in addition to the use of `const` where appropriate).

nwiltsie replied:
Done with ddf79e9.

aholmes commented on .github/workflows/nextflow-tests.yml:
```diff
@@ -73,11 +169,257 @@ jobs:
+            // https://docs.github.com/en/rest/guides/using-the-rest-api-to-interact-with-your-git-database
+            for (let i=0; i<10; i++) {
+              // Wait the first time - it will always take some time
+              await timers.setTimeout(3000)
```
Do we wait the first time because the auto-commit takes time to wrap up?

nwiltsie replied:
It's not that auto-commit takes time to wrap up (it should have fully completed by the time this step started), it's that GitHub takes time to compute the mergeability and update the branch. Per the docs:

> A consuming API needs to explicitly request a pull request to create a test merge commit. A test merge commit is created when you view the pull request in the UI and the "Merge" button is displayed, or when you [get](https://docs.github.com/en/rest/pulls/pulls#get-a-pull-request), [create](https://docs.github.com/en/rest/pulls/pulls#create-a-pull-request), or [edit](https://docs.github.com/en/rest/pulls#update-a-pull-request) a pull request using the REST API. Without this request, the merge Git refs will fall out of date until the next time someone views the pull request.

Without waiting, the request creation fails with something like a "no such commit" error (I can't find an example). This loop just keeps checking until those things have updated.

aholmes replied:
Got it. I'm not sure if I'm misreading something - 

This block is the PR request to create the test merge commit, right?

```javascript
var updated_pr = (await github.rest.pulls.get({
  owner: context.repo.owner,
  repo: context.repo.repo,
  pull_number: ${{ toJSON(needs.discover.outputs.pr_number) }}
})).data
```

If that's correct, why do we wait before the request?

nwiltsie replied:
I... can't defend that, honestly. I feel like I had a reason, but by my own logic we shouldn't. Huh! Okay, I'll move the wait to the end of the block.

aholmes replied:
I'm sure we will rediscover the reason by surprise some time in the future. :)

nwiltsie replied:
Done with 8393a19.

aholmes commented on .github/workflows/nextflow-tests.yml:
```diff
@@ -73,11 +169,257 @@ jobs:
+              if ((updated_pr.head.sha != ${{ toJSON(needs.discover.outputs.review_sha) }})
+                  &&
+                  (updated_pr.mergeable != null)) {
+                break
```
Should this handle the case of hitting 10 loops and never succeeding?

nwiltsie replied:
Eh... the only time I've ever seen it take more than 3 seconds is when I get the "no such commit" error above because I made some mistake. I don't know what else to do other than fail if it does take more than 30 seconds.

aholmes replied:
Got it - this means `await github.rest.pulls.createReview(review_details)` will fail anyway? That's fine.

nwiltsie replied:
Correct - all of the waiting is just trying to ensure that `createReview` won't fail, so in the worst case it does just fail anyway.

aholmes commented on run-nextflow-tests/configtest.py:
```diff
@@ -18,6 +18,9 @@
 T = TypeVar('T', bound='NextflowConfigTest')
 
 
+BAD_CHARACTERS = re.compile(r'[":<>|*?\r\n\\/]')
```
I suggest whitelisting and restricting to a list of characters we know are safe.

nwiltsie replied:
This list comes from the upload-artifacts action: https://github.com/actions/upload-artifact/issues/333

> Invalid characters include:  Double quote ", Colon :, Less than <, Greater than >, Vertical bar |, Asterisk *, Question mark ?, Carriage return \r, Line feed \n

I didn't open that issue, but I do recall getting a similar warning. I also added `/` as a disallowed character.

aholmes replied:
Okay. I'm okay to leave it for now, though generally having a whitelist is safer as it doesn't require chasing down all possible problematic characters by just blanket disallowing anything not expected.

aholmes commented on run-nextflow-tests/configtest.py:
```diff
@@ -255,27 +258,67 @@ def print_diffs(self, other: T):
+                    json.dump(prior.expected_result, outfile)
+
+                with after.open(mode="w", encoding="utf-8") as outfile:
+                    json.dump(self.expected_result, outfile)
```
Where does `self.expected_result` actually get set?

I'm having a little trouble following the exact flow here.

I see that we have these two concepts of "a prior expected result" and some "expected result" that we write to "before" and "after" (respectively). The prior result comes from wherever called this method, but the other expected result is a class member and set "globally"?

Then we run `jd` and add a collection of EOF blocks to a file.

Perhaps the class-member `expected_result` cannot be changed in this PR (out of scope). I think some minor changes to variable names in this method, and perhaps splitting this method into a couple different methods might clarify what's happening. Some possible candidates:

- `outfile` is redefined four times in this method; without scanning the entire method to see when it changes, it's easy to hold the wrong file in mind when reading the code.
- What are "before" and "after"? I think, knowing what this PR is about, we're looking at some PR content "before" any automation ran, and then PR content "after"?
- Why does this only run if `not test_passed`? Could this be presented as something like `write_pr_failure_notes` or some such? I think this method doesn't "know" what passing or failing tests are, but the caller of it does, so allowing the caller to decide whether to "write PR notes" based on its knowledge of failing tests may clarify the logic of whether to run this block.
- Because there are two sort of "modes" here - writing more data if `not test_passed`, and writing data either way, splitting the `not test_passed` block into a separate method may simplify reading this method.

Lastly, is there somewhere obvious I missed that explains what the EOF blocks are for? Maybe a short comment directing to information on what they are can help.

nwiltsie replied:
Firstly, the `EOF` nonsense is used to [pass multiline-outputs](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#multiline-strings) to `GITHUB_OUTPUT`. The `eof_index` chicanery is to make sure that filenames with `EOF` in them don't cause problems.

nwiltsie replied:
At a high level, here is the general flow:

1. A `NextflowConfigTest` (a dataclass) is [parsed from a file](https://github.com/uclahs-cds/tool-Nextflow-action/blob/43877d5401e3240e80863a3d4d8bd400a8c0f86f/run-nextflow-tests/entry.py#L13)
2. `recompute_results()` [generates a new dataclass instance](https://github.com/uclahs-cds/tool-Nextflow-action/blob/43877d5401e3240e80863a3d4d8bd400a8c0f86f/run-nextflow-tests/entry.py#L18) by running the tests. [Internally](https://github.com/uclahs-cds/tool-Nextflow-action/blob/43877d5401e3240e80863a3d4d8bd400a8c0f86f/run-nextflow-tests/configtest.py#L323-L339) that does a [little bit more](https://github.com/uclahs-cds/tool-Nextflow-action/blob/43877d5401e3240e80863a3d4d8bd400a8c0f86f/run-nextflow-tests/configtest.py#L63-L78) than `dataclasses.replace()`. _This is where `expected_results` gets updated_.

From here it gets a little more complicated, I'll handle that in a separate comment.

nwiltsie replied:
I reworked a lot of this with 0bcfd6a (and 43f7448) with a more streamlined flow and more comments. Does it look better to you now?

nwiltsie replied:
Ahh, and [that's](https://github.com/uclahs-cds/tool-Nextflow-action/actions/runs/8525747220/job/23353433322) why I re-used `outfile` so much - that method is right on the edge of pylint's [too-many-locals](https://pylint.readthedocs.io/en/latest/user_guide/messages/refactor/too-many-locals.html) limit.

aholmes replied:
> ... that method is right on the edge of pylint's [too-many-locals](https://pylint.readthedocs.io/en/latest/user_guide/messages/refactor/too-many-locals.html) limit.

Ah, interesting! I think this is a good example of a linter's limitations and when their "objective" nature starts to tread into subjectiveness. Short of endless heuristics, how can it decide whether it's better to reuse a local to stay under that limit, or create a new one to increase understanding? How can it prescribe what "too many locals" is for _every_ method? I have my own opinions about what's "best" as I'm sure we all do. ¯\\\_(ツ)_/¯

aholmes replied:
> I reworked a lot of this with [0bcfd6a](https://github.com/uclahs-cds/tool-Nextflow-action/commit/0bcfd6ae562fbf78a46269888af055d91ab667dc) (and [43f7448](https://github.com/uclahs-cds/tool-Nextflow-action/commit/43f7448487315835dce31ae624df2310c44bc938)) with a more streamlined flow and more comments. Does it look better to you now?

Yeah, looks great! It's quite clear what's happening, and both the `print_only` variable name and its explanation bring in context that I wasn't previously aware of, so that's awesome.

I just have one comment I added.

nwiltsie commented:
Okay, I believe I've addressed all comments and run the process through its paces with https://github.com/uclahs-cds/user-nwiltsie-pipeline/pull/12.

(Technically I was using a different branch so I could use a test image, but [they're nearly identical](https://github.com/uclahs-cds/tool-Nextflow-action/compare/nwiltsie-autofix-tests..nwiltsie-slash-command-workflow).)
